source/lac/slepc_solver.cc

```diff
@@ -195,22 +195,23 @@ namespace SLEPcWrappers
       for (unsigned int i = 0; i < *n_converged; i++)
         {
           double residual_norm_i = 0.0;
-          // EPSComputeResidualNorm is L2-norm and is not consistent with the stopping criteria
-          // used during the solution process.
-          // Yet, this is the norm which gives error bounds (Saad, 1992, ch3):
-          //   | \lambda - \widehat\lambda | <= ||r||_2
-          ierr = EPSComputeResidualNorm (eps, i, &residual_norm_i);
-
-          // EPSComputeRelativeError may not be consistent with the stopping criteria
-          // used during the solution process. Given EPS_CONV_ABS set above,
-          // this can be either the l2 norm or the mass-matrix induced norm
-          // when EPS_GHEP is set.
-          // ierr = EPSComputeRelativeError (solver_data->eps, i, &residual_norm_i);
-
-          // EPSGetErrorEstimate is consistent with the residual norm
-          // used during the solution process. However, it is not guaranteed to
-          // be derived from the residual even when EPSSetTrueResidual is set.
-          // ierr = EPSGetErrorEstimate (solver_data->eps, i, &residual_norm_i);
+          // EPSComputeError (or, in older versions of SLEPc,
```

Comment:
Have a look at this thread https://lists.mcs.anl.gov/pipermail/petsc-users/2014-November/023509.html

`EPSComputeResidualNorm` always use L2 norm but it's not necessarily what SLEPC checks internally. 
`EPSComputeRelativeError` is L2 for non-symm and mass-matrix induced norm for GHEP. 
From that thread it seems `EPSSetTrueResidual` is not implemented in all solvers, so you kind of never know.

p.s. I must admit I have not used SLEPc for a while

Reply:
@davydden good point. I updated the comments to include a link to that thread so our choices in the code are clear :+1: